### DISCUSSION TITLE
Abstract Polymorphism

### DISCUSSION THREAD

**Discussion by SlimyFrog123** [sqlalchemy/sqlalchemy#12161] [Usage Questions]
I'm still fairly new to SQLAlchemy, so I'm not sure what I'm doing wrong, and nothing I found online seemed to help.

I'm using Flask-SQLAlchemy. I have my `Base` model (which everything inherits):
```py
class Base(db.Model):
    __abstract__ = True

    id = Column(Integer, primary_key=True)
    created_at = Column(DateTime, default=utcnow)
    updated_at = Column(DateTime, default=utcnow, onupdate=utcnow)
```
Then I have my `Namespace` model:
```py
class Namespace(Base):
    __tablename__ = 'namespaces'

    namespace = Column(String(255), unique=True, nullable=False)
```
Then I have a `NamespaceItem` model (self-explanatory):
```py
class NamespaceItem(Base):
    __tablename__ = 'namespace_items'
    __abstract__ = True

    namespace_id = Column(Integer, ForeignKey('namespaces.id'), nullable=False)
    identifier = Column(String(255), nullable=False)
``` 
Now, I have a couple tables (more in the future) which extend the `NamespaceItem` model:
```py
class Group(NamespaceItem):
    __tablename__ = 'groups'

    name = Column(String(128), unique=True, nullable=False)

class Widget(NamespaceItem):
    __tablename__ = 'widgets'

    source = Column(Integer, ForeignKey('widget_sources.id'), nullable=False)
    content = Column(Text, nullable=False)
```

I need to check if an item exists in the namespace (e.g. `example_namespace.test_item`), so I try to use a query:
```py
db_session.query(NamespaceItem).filter_by(namespace_id=namespace_instance.id, identifier=identifier).first()
```
Which throws me the error: `sqlalchemy.exc.ArgumentError: Column expression, FROM clause, or other columns clause element expected, got <class '__main__.NamespaceItem'>.`
That's fine. I did some research and came across the `with_polymorphic` method, so I tried:
```py
namespace_item = with_polymorphic(NamespaceItem, '*') # I tried [Group, Widget] in place of '*', but it changed nothing.
item = db_session.query(namespace_item).filter_by(namespace_id=namespace_instance.id, identifier=identifier).first()
```
That gives me the error: `sqlalchemy.orm.exc.UnmappedClassError: Class '__main__.NamespaceItem' is not mapped`

I don't want to have `NamespaceItem` as a table in the database since it's strictly for inheritance purposes, but I tried it anyway by changing it from abstract and adding a `__tablename__`, but then it just gave me an error saying that another class of mine, `GroupMembership`:
```py
class GroupMembership(Base):
    __tablename__ = 'group_memberships'

    user_id = Column(Integer, ForeignKey('users.id'), nullable=False)
    group_id = Column(Integer, ForeignKey('groups.id'), nullable=False)
```
...was trying to access the key `groups.id` when `groups` doesn't exist...even though groups is created before it, so I'm not sure what to do now.

**Comment by CaselIT** [comment]
Hi,

I would suggest you refer to the docs: https://docs.sqlalchemy.org/en/20/orm/inheritance.html

To define an inheritance relationship you need to configure sqlalchemy correctly
